### PR TITLE
feat(pt): log pre-clip gradient total_norm and per-param norm to tensorboard

### PR DESCRIPTION
**Pull request by OutisLi**
## Summary by CodeRabbit

* **New Features**
  * Enhanced gradient diagnostics during training: when TensorBoard logging is enabled, training now records total gradient norm, per-parameter gradient norms, and a histogram of per-parameter norms. The system also highlights the top-10 largest parameter gradients as individual scalars and surfaces total-norm values in training logs to improve analysis and monitoring.

### CODE REVIEW

File: deepmd/pt/train/training.py [hunk 1]
```diff
@@ -990,7 +990,20 @@ def step(_step_id: int, task_key: str = "Default") -> None:
...
                 loss.backward()
+                # === Initialize gradient diagnostics variables ===
+                total_norm: torch.Tensor | None = None
+                pre_clip_named_norms: list[tuple[str, float]] = []
                 if self.gradient_max_norm > 0.0:
+                    # Collect per-parameter gradient norms before clipping.
+                    # NOTE: Under FSDP2 with ZeRO stage >= 2, p.grad is a sharded DTensor,
+                    # so p.grad.norm() computes the shard-local L2 norm, not the full-parameter
+                    # norm. Skip per-param collection in this case to avoid misleading values.
+                    if self.enable_tensorboard and self.zero_stage < 2:
```
Comment (njzjz) [inline deepmd/pt/train/training.py]: `tensorboard_freq` should be considered, since the D2H operator below is not cheap.

File: deepmd/pt/train/training.py [hunk 2]
```diff
@@ -1342,6 +1355,34 @@ def log_loss_valid(_task_key: str = "Default") -> dict:
...
+                    writer.add_scalar(
+                        f"{task_key}/grad/total_norm",
+                        total_norm.item(),
+                        display_step_id,
+                    )
+                    # Only log per-parameter norms if list is non-empty.
+                    if pre_clip_named_norms:
+                        norms = torch.tensor(
+                            [gn for _, gn in pre_clip_named_norms],
+                            dtype=torch.float32,
```
Comment (njzjz) [inline deepmd/pt/train/training.py:1447]: why is it cast to float32?

File: deepmd/pt/train/training.py [hunk 3]
```diff
@@ -1342,6 +1355,34 @@ def log_loss_valid(_task_key: str = "Default") -> dict:
...
+                        writer.add_histogram(
+                            f"{task_key}/grad/param_norms", norms, display_step_id
+                        )
+                        # Log top-10 largest per-parameter gradient norms.
+                        # Shorten name: keep everything after "atomic_model.".
+                        pre_clip_named_norms.sort(key=lambda x: x[1], reverse=True)
+                        for name, gn in pre_clip_named_norms[:10]:
+                            idx = name.find("atomic_model.")
+                            if idx >= 0:
+                                name = name[idx + len("atomic_model.") :]
```
Comment (njzjz) [inline deepmd/pt/train/training.py]: This may cause a name conflict for variables with different prefixes. In the design, we never force a single atomic model.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/pt/train/training.py (hunk 2 above)
Comment (OutisLi) [reply deepmd/pt/train/training.py:1447]: since the linter require explicit dtype, and fp32 is a safe choice

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]
